programs/ssl/ssl_client2.c

```diff
@@ -543,7 +544,7 @@ struct options {
     int reproducible;           /* make communication reproducible          */
     int skip_close_notify;      /* skip sending the close_notify alert      */
 #if defined(MBEDTLS_SSL_EARLY_DATA)
-    int early_data;             /* support for early data                   */
+    const char *early_data_file; /* the file path of early data             */
```

Comment:
> Should we keep `early_data` parameter?
> How about the behaviors when early data is enable and `mbedtls_ssl_write_early_data()` is NOT called?Should/does client send `early_data` extension in this case?


--------------------
In my opinion, if we don't provide one valid input early data file, we should not enable early data.  That means:  in this design, there is no such way to enable early data but don't call mbedtls_ssl_write_early_data().  Anyway, we need talk about whether the design is reasonable and get one final decision. @ronald-cron-arm @yuhaoth @tom-cosgrove-arm

Reply:
>if we don't provide one valid input early data file, we should not enable early data. That means: in this design, there is no such way to enable early data but don't call mbedtls_ssl_write_early_data().

That's design of `ssl_server2`. I just want to make sure the behavior of above case and should we cover it in `ssl_client2`

Reply:
> In my opinion, if we don't provide one valid input early data file, we should not enable early data. That means: in this design, there is no such way to enable early data but don't call mbedtls_ssl_write_early_data(). Anyway, we need talk about whether the design is reasonable and get one final decision. @ronald-cron-arm @yuhaoth @tom-cosgrove-arm

I think we need to test the edge but valid case where the client claims that it is going to send early data but it actually does not send any.

If I understand correctly, two possible approaches are proposed here for the early data ssl_client2 option(s):
1. only one  ssl_client2 option, early_data_file: if the option is not set early data is not enabled, if the option is set but the file is empty, early data is enabled but no early data send ...
2. two ssl_client2 options, one to enable and the other to pass the path of the file containing the data to transfer. In that case no file or an empty file are equivalent.

Both work but the second one is better practice I would say (but I can be wrong), separation of purpose.

Reply:
I prefer second one with some modifications. 
- `early_data` : reserve, enable/disable early data
- `early_data_file`: path of early data file. When set, `mbedtls_ssl_write_handshake()` should be called in next PR. Otherwise, `mbedtls_ssl_handshake()` will be called. 
In next PR, we can make final decision which one should be tested. 

This PR is for adding options. I think we should not limit the possible cases in this PR.

Reply:
Thank you for your suggestions, will follow up the second way.

`2 .two ssl_client2 options, one to enable and the other to pass the path of the file containing the data to transfer. In that case no file or an empty file are equivalent.`